## Question 1 : Dimension reduction in machine learning
#### Dimensionalty reduction is the process of reducing the dimension of the feature set. This process is important when the data consists huge no of features and it will lead to the problem of curse of dimentionality. There are different ways to reduce the dimensions such as:
#### 1. PCA(principle component analysis)
#### 2. Factor analysis
#### 3. Univariate selection
#### 4. Variance thresholding etc.

### Question 2 : Identify duplicates and remove them

In [37]:
# importing pandas library for reading the data into a notebook.
import pandas as pd

In [38]:
# Reading the data from csv file
df = pd.read_csv('house data.csv')

In [39]:
df.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [40]:
# Using this commad we can easily identify the no of rows and cols in the data
df.shape

(13320, 9)

### In this dataset we have multiple values of locations.
### Now suppose we only want the unique names of locations.
### So first we need to identify the duplicate values from location column
### Then we will drop the entire row of the duplicate values.

In [41]:
# There are 12014 duplicate values in the location col.
df['location'].duplicated().sum()

12014

In [42]:
df[df['location'].duplicated()].head()

area_type   availability         location       size  society  \
10  Super built-up  Area         18-Feb       Whitefield      3 BHK      NaN   
11            Plot  Area  Ready To Move       Whitefield  4 Bedroom  Prrry M   
27        Built-up  Area         20-Dec       Whitefield      3 BHK      NaN   
29  Super built-up  Area  Ready To Move  Electronic City      3 BHK  KBityo    
31  Super built-up  Area  Ready To Move    Bisuvanahalli      3 BHK  Prityel   

   total_sqft  bath  balcony  price  
10       1800   2.0      2.0   70.0  
11       2785   5.0      3.0  295.0  
27       1610   3.0      2.0   81.0  
29       1025   2.0      1.0   47.0  
31       1075   2.0      1.0   35.0

In [43]:
# Here we are dropping the entire row of the duplicate values in the location column.
df.drop_duplicates(subset=['location']).head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [44]:
df.drop_duplicates(subset=['location']).shape

(1306, 9)

### Conclusion:
####             Here we have dropped the 1306 rows which corresponds to the duplicate values of location column.